draft-jhoyla-req-mtls-flag.md

```diff
@@ -32,48 +32,99 @@ informative:
+MUST NOT prompt the user or take actions that would prompt the user for example
+trying to read certificates from off-board devices such as smart cards.
 
 # Conventions and Definitions
 
 {::boilerplate bcp14-tagged}
 
 # Flag specification
 
-A server receiving this flag MAY send a CertificateRequest message.
+The `request_client_auth` flag is sent by the client in the `ClientHello`
+message, and, if accepted by the server, acknowledged in the server's
+`CertificateRequest` message. A server that receives this flag and wishes to
+accept the extension MUST send a `CertificateRequest` message with the TLS
+Flags extension, and with the `request_client_auth` flag set. The server MAY
+send a `CertificateRequest` message without the `request_client_auth` flag set
+if it wishes to negotiate client authentication for another reason. This could
+occur if the server is configured to always request a client certificate. The
+client sending this flag MUST send a `Certificate` message if it receives a
+`CertificateRequest` with the `request_client_auth` flag set, even if said
+`Certificate` message is empty, and SHOULD NOT preemptively terminate the
+connection anticipating a "certificate_required" alert. Appropriate cases where
+a client would preemptively terminate the connection include where the
+`CertificateRequest` includes a `certificate_authorities` extension or
+`signature_algorithms` extension that the client cannot satisfy.
+
+A server receiving the `request_client_auth` flag in a `ClientHello` message
+that wishes to accept the flag MUST echo the `request_client_auth` flag in the
+`CertificateRequest` message. This informs the client that the reason for the
+`CertificateRequest` was because of the `request_client_auth` flag and not due
+to some other policy. This allows clients to select certificates based on
+whether the server accepted the flag or not.
+
+A server MUST NOT set the `request_client_auth` flag in the
+`CertificateRequest` unless it received the flag in the `ClientHello`. A client
+receiving the `request_client_auth` flag that did not set it in the
+`ClientHello` MUST generate a fatal `illegal_parameter` alert.
+
+A client that receives this flag in any message other than `CertificateRequest`
+MUST terminate the connection with a fatal `illegal_parameter` alert. Similarly
+a server that receives this flag in any message other than `ClientHello` MUST
+terminate the connection with a fatal `illegal_parameter` alert.
 
 # Security Considerations
 
-This flag should have no effect on the security of TLS, as the server may always
-send a CertificateRequest message during the handshake. This flag merely
-provides a hint that the client will be able to fulfil the request. If the
-client sets this flag but then fails to provide a certificate the server MAY
-terminate the connection with a bad_certificate error.
+This flag should have no effect on the security of TLS, as the server may
```

Comment:
This is interesting because even if the flag on its own has no security impact, the use case we intend to support here -- web bot auth -- certainly does. Namely, we have to override the logic the server uses to decide whether to send a CertificateRequest, how to select a trust anchor to validate the Certificate, and how to handle application data. We don't want to authorize HTTP requests from a crawler that would otherwise require a different client identity.

Given the WG feedback so far, I think we'll need to discuss this here. If you'd rather do it in a separate PR, then feel free to just leave a TODO here.

Reply:
Filed https://github.com/jhoyla/draft-jhoyla-req-mtls-flag/issues/6.